packages/server-wallet/src/protocols/channel-opener.ts

```diff
@@ -38,34 +38,61 @@ export class ChannelOpener {
       if (!channel) {
         throw new Error(`ChannelOpener can't find channel with id ${channelToLock}`);
       }
+      // always queue the channel if we've potentially touched it
+      response.queueChannel(channel);
 
       // if we haven't signed the pre-fund, sign it
       if (!channel.prefundSigned) {
         await this.signPrefundSetup(channel, response, tx);
       }
 
-      // if we don't have a complete preFundSetup, we're done for now
+      // We might be waitingFor theirPreFundState
+      let waitingFor = 'theirPreFundState';
+
+      // if we don't have a complete preFundSetup, we are still waitingFor theirPreFundState and cannot progress
       if (!channel.preFundComplete) {
-        response.queueChannel(channel); // always queue the channel if we've potentially touched it
+        await updateWaitingFor();
         return;
       }
 
+      // otherwise, we might be waiting on funding
+      waitingFor = 'funding';
+
       // if we have a full pre fund setup, delegate to the funding process to (a) see if
       // the channel is funded, and (b) take action if not
       const funded = await this.crankChannelFunder(objective, channel, response, tx);
 
+      // if the channel is not funded, we are still waitingFor funding and cannot progress
+      if (!funded) {
+        await updateWaitingFor();
```

Comment:
According to my understanding, I don't think this pattern will work well in practice.

It seems like, when we hit L67, we will `await updateWaitingFor()` every time we call `crank`. That will update the timestamp to now. IE. we could repeatedly call `crank`, and update the timestamp to `now()`, without the state changing.

This means that the timestamp is not actually tracking the last-known-state-change.

Reply:
If you look at the implementation of `updateWaitingFor()`, you'll notice that it calls the function that will update the timestamp *only if* `waitingFor` has changed. 

The outer function should probably be named `updateWaitingForIfChanged()`.